maths/binary_exponentiation.py

```diff
@@ -1,48 +1,148 @@
-"""Binary Exponentiation."""
+"""
+Binary Exponentiation
 
-# Author : Junth Basnet
-# Time Complexity : O(logn)
+This is a method to find a^b in O(log b) time complexity and is one of the most commonly
+used methods of exponentiation. The method is also useful for modular exponentiation,
+when the solution to (a^b) % c is required.
 
+To calculate a^b:
+- If b is even, then a b = (a * a)^(b / 2)
+- If b is odd, then a^b = a * a^(b - 1)
+Repeat until b = 1 or b = 0
 
-def binary_exponentiation(a: int, n: int) -> int:
+For modular exponentiation, we use the fact that (a * b) % c = ((a % c) * (b % c)) % c
+"""
+
+
+def binary_exp_recursive(base: int, exponent: int) -> int:
     """
-    Compute a number raised by some quantity
-    >>> binary_exponentiation(-1, 3)
+    >>> binary_exp_recursive(3, 5)
```

Comment:
What about tests for a=big number andn=big number?

Reply:
Can do, but the n=float case won't work because the algorithm can only compute integer powers

Reply:
I know it won’t work but I wanted to see how it fails.  Does it raise a ValueError?  Does it behave like `pow()` does.  Testing how things work if half the battle.  Watching them fail is just as cool.